### PR TITLE
cleanup(libsinsp): extend ppm_sc sinsp API

### PR DESCRIPTION
**Pull request by incertum**
Signed-off-by: Melissa Kilby <melissa.kilby.oss@gmail.com>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

Expose additional sinsp APIs around configuring interesting sets of syscalls and events. 
https://github.com/falcosecurity/falco/pull/2361

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
cleanup(libsinsp): extend ppm_sc sinsp API
```

### CODE REVIEW

File: userspace/libsinsp/sinsp_ppm_sc.cpp [hunk 1]
```diff
@@ -238,6 +251,57 @@ std::unordered_set<uint32_t> sinsp::enforce_sys_ppm_sc_set(std::unordered_set<ui
 	return ppm_sc_set;
 }
 
+std::unordered_set<std::string> erase_io_syscalls_ppm_sc_codes(std::unordered_set<uint32_t>& ppm_sc_of_interest, std::unordered_set<uint32_t>& io_ppm_sc_set)
```
Comment (incertum) [inline userspace/libsinsp/sinsp_ppm_sc.cpp:254]: @FedeDP would you have more ideas here? It doesn't fit the current sinsp API style, but for Falco we would also need the set of erased syscalls names. Happy to split and adjust depending on what you would prefer the most. Thanks!

File: userspace/libsinsp/sinsp_ppm_sc.cpp [hunk 2]
```diff
@@ -238,6 +251,57 @@ std::unordered_set<uint32_t> sinsp::enforce_sys_ppm_sc_set(std::unordered_set<ui
...
+			if (sinsp::is_metaevent(ev) || ev == PPME_PROCEXIT_1_E)
+			{
+				ppm_event_info_of_interest.insert(ev);
+			}
+		}
+	}
+	return ppm_event_info_of_interest;
+}
+
+std::unordered_set<uint32_t> get_syscalls_ppm_sc_codes(const std::unordered_set<std::string>& syscalls_names)
```
Comment (FedeDP) [inline userspace/libsinsp/sinsp_ppm_sc.cpp:291]: Yep, we missed this conversion utility!

File: userspace/libsinsp/sinsp_ppm_sc.cpp [hunk 3]
```diff
@@ -238,6 +251,57 @@ std::unordered_set<uint32_t> sinsp::enforce_sys_ppm_sc_set(std::unordered_set<ui
...
+		if (io_ppm_sc_set.find(ppm_sc_code) != io_ppm_sc_set.end())
+		{
+			ppm_sc_of_interest.erase(ppm_sc_code);
+			erased_io_syscalls_names.insert(g_infotables.m_syscall_info_table[ppm_sc_code].name);
+		}
+	}
+	return erased_io_syscalls_names;
+}
+
+std::unordered_set<uint32_t> enforce_events(std::unordered_set<uint32_t> ppm_event_info_of_interest)
```
Comment (FedeDP) [inline userspace/libsinsp/sinsp_ppm_sc.cpp:270]: Mmmh i think that non-syscalls events are always enabled, right?
Why do we need this?

File: userspace/libsinsp/sinsp_ppm_sc.cpp (hunk 1 above)
Comment (jasondellaluce) [reply userspace/libsinsp/sinsp_ppm_sc.cpp:254]: Intersection, union, and difference operations on `unordered_set`s are starting to become common ops both here and in Falco. I'd be in favor of creating generig utilities/helpers for those as part of the libsinsp utils eventually.

**Comment by incertum** [comment]
ACK -> intersection, union, and difference operations on unordered_sets, convert from unordered to ordered (btw already exists in libs utils), convert from vectors to sets etc ... I'll implement all these utils here in libs in a follow up PR and then we don't need them in Falco as I already started adding some of them in Falco in the other PR -> hence fully agree with having everything in one place in libs utils.

In summary new additions now:

`get_ppm_sc_set_from_syscalls`
`enforce_events` -> yes we need this here, it's currently that way in Falco and sinsp API is a better place also with unit testing across repos in mind.

And yes we can get away with combinations of util functions for the rest.
Thanks @FedeDP and @jasondellaluce, please take another look now.

**Comment by FedeDP** [comment]
> hence fully agree with having everything in one place in libs utils.

Agree! Thank also @jasondellaluce for your input!

File: userspace/libsinsp/sinsp_ppm_sc.cpp [hunk 4]
```diff
@@ -238,6 +251,41 @@ std::unordered_set<uint32_t> sinsp::enforce_sys_ppm_sc_set(std::unordered_set<ui
 	return ppm_sc_set;
 }
 
+std::unordered_set<uint32_t> sinsp::enforce_events(std::unordered_set<uint32_t> ppm_event_info_of_interest)
```
Comment (FedeDP) [inline userspace/libsinsp/sinsp_ppm_sc.cpp:254]: I see now; perhaps we can call this one (following other existing method) `enforce_sinsp_state_ppme`?

**Comment by FedeDP** [comment]
(i think you need to rebase if you want to remove the `test-libraries-arm64` failure! Even if it is not a required job though ;) )

**Comment by incertum** [comment]
Thanks, done ;)

File: userspace/libsinsp/sinsp_ppm_sc.cpp (hunk 1 above)
Comment (Andreagit97) [reply userspace/libsinsp/sinsp_ppm_sc.cpp]: why not use directly `std::set` in these APIs? we are at initialization time so probably also `std::set` should be ok :thinking:

Comment (incertum) [reply userspace/libsinsp/sinsp_ppm_sc.cpp]: I think at this point we are stuck with unordered_set and since we need to land the adaptive syscalls selection mechanisms (spanning multiple PRs across both repos), it could be something to be considered for a refactor later? WDYT? In general yes I agree, just the timing seems not right atm.

Comment (Andreagit97) [reply userspace/libsinsp/sinsp_ppm_sc.cpp]: make sense, i would only avoid to introduce new helpers for unordered sets if we don't need them :)

**Review by FedeDP — APPROVED** [review]
/approve
Thank you!

**Review by jasondellaluce — APPROVED** [review]
